3-enrich/beam-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich.beam/utils.scala

```diff
@@ -0,0 +1,167 @@
+        ee.contexts = write(getContextParentEvent(ee.event_id))
+        ee
+      }
+
+  private def getContextParentEvent(eventId: String): JValue =
+    ("schema" -> "iglu:com.snowplowanalytics.snowplow/contexts/jsonschema/1-0-0") ~
+      ("data" -> List(
+        ("schema" -> "iglu:com.snowplowanalytics.snowplow/parent_event/jsonschema/1-0-0") ~
+          ("data" -> ("parentEventId" -> eventId))))
+
+  def emitPii(registry: EnrichmentRegistry): Boolean =
+    registry.getPiiPseudonymizerEnrichment.map(_.emitIdentificationEvent).getOrElse(false)
+
+  /**
+   * Truncate an oversized formatted enriched event into a [[BadRow]].
+   * @param enrichedEvent TSV-formatted oversized enriched event
+   * @param bytesSize size in bytes of the formatted enriched event
+   * @param maxBytesSize maximum size in bytes a record can take
+   * @return a [[BadRow]] containing a the truncated enriched event (10 times less than the max size)
+   */
+  def resizeEnrichedEvent(enrichedEvent: String, bytesSize: Int, maxBytesSize: Int): BadRow =
+    BadRow(enrichedEvent.take(maxBytesSize / (4 * 10)), NonEmptyList(
+      s"Size of enriched event ($bytesSize) is greater than allowed maximum ($maxBytesSize)"))
+
+  /**
+   * Resize a [[BadRow]] if it exceeds the maximum allowed size.
+   * @param badRow the original [[BadRow]] which can be oversized
+   * @param maxBytesSize maximum size in bytes a record can take
+   * @return a [[BadRow]] where the line is 10 times less than the max size
+   */
+  def resizeBadRow(badRow: BadRow, maxBytesSize: Int): BadRow = {
+    val size = getStringSize(badRow.line)
+    if (size >= maxBytesSize) {
+      val msg = s"Size of bad row ($size) is greater than allowed maximum size ($maxBytesSize)"
+      badRow.copy(
+        line = badRow.line.take(maxBytesSize / (4 * 10)),
+        errors = msg.toProcessingMessage <:: badRow.errors
+      )
+    } else badRow
+  }
+
+  /** Get the size of a string in bytes. */
+  def getStringSize(string: String): Int =
+    ByteBuffer.wrap(string.getBytes(UTF_8)).capacity
+
+  /** Measure the time spent in a block of code in milliseconds. */
+  def timeMs[A](call: => A): (A, Long) = {
+    val t0 = System.currentTimeMillis()
+    val result = call
+    val t1 = System.currentTimeMillis()
+    (result, t1 - t0)
+  }
+
+  /**
+   * Create a symbolic link.
+   * @param file to create the sym link for
+   * @param symLink path to the symbolic link to be created
+   * @return either the path of the created sym link or the error
+   */
+  def createSymLink(file: File, symLink: String): Either[String, Path] = {
+    val symLinkPath = Paths.get(symLink)
+    if (Files.notExists(symLinkPath)) {
+      Try(Files.createSymbolicLink(symLinkPath, file.toPath)) match {
+        case scala.util.Success(p) => Right(p)
+        case scala.util.Failure(t) => Left(s"Symlink can't be created: ${t.getMessage}")
+      }
+    } else Left(s"Symlink $symLinkPath already exists")
```

Comment:
This entry may exist but that does not mean it is a symlink. Also as per documentation notExists: "method is not the complement of the exists method. Where it is not possible to determine if a file exists or not then both methods return false."

Reply:
good point, I added a comment to the constant saying: "We want to take one-tenth of the payload characters and one character can take up to 4 bytes"

Reply:
woops, wrong comment